modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2891,7 +2891,15 @@ def combine(self) -> PandasDataframe:
         PandasDataframe
             A single partition PandasDataframe.
         """
-        partitions = self._partition_mgr_cls.combine(self._partitions)
+        new_index = None
+        new_columns = None
+        if self._deferred_index:
+            new_index = self.index
+        if self._deferred_column:
+            new_columns = self.columns
```

Comment:
A further improvement could be to get rid of the materialization of indexes in the main process. However this also happens in `_propagate_index_objs` function now.